pkg/resources/reconciler.go

```diff
@@ -228,47 +228,82 @@ func (r *Reconciler) ReconcileSubscription(ctx context.Context, target marketpla
+		var csv *operatorsv1alpha1.ClusterServiceVersion
+		if sub.Status.InstalledCSV != "" {
+			csv = &operatorsv1alpha1.ClusterServiceVersion{
+				ObjectMeta: metav1.ObjectMeta{Namespace: target.Namespace, Name: sub.Status.InstalledCSV},
 			}
-			log.Warningf("Deleting subscription for re-install due to failed install plan", l.Fields{"ns": target.Namespace, "install plan": target.Pkg})
-			if err := client.Delete(ctx, sub); err != nil && !k8serr.IsNotFound(err) {
-				return integreatlyv1alpha1.PhaseFailed, fmt.Errorf("failed to delete subscription for re-intstall: %s", err)
+		}
+		return retryInstallation(ctx, client, log, target, csv, sub)
+	}
+
+	//if it's approved but not complete, then it's in progress
+	if ip.Status.Phase != operatorsv1alpha1.InstallPlanPhaseComplete && ip.Spec.Approved {
+		log.Infof("Install plan is not complete yet ", l.Fields{"install plan": target.Pkg})
+		return integreatlyv1alpha1.PhaseInProgress, nil
+		//if it's not approved by now, then it will not be approved by this version of the integreatly-operator
+	} else if !ip.Spec.Approved {
+		log.Infof("Upgrade installplan above the maximum allowed version", l.Fields{"install plan": target.Pkg})
+	}
+
+	for _, csvName := range ip.Spec.ClusterServiceVersionNames {
+		ipCSV := &operatorsv1alpha1.ClusterServiceVersion{}
+		if err := client.Get(ctx, k8sclient.ObjectKey{
+			Name:      csvName,
+			Namespace: target.Namespace,
+		}, ipCSV); err != nil {
+			if k8serr.IsNotFound(err) {
+				log.Infof("Waiting for CSV to be created in cluster after InstallPlan is complete: %s", l.Fields{
+					"install plan": target.Pkg,
+				})
+				return integreatlyv1alpha1.PhaseInProgress, nil
 			}
-			return integreatlyv1alpha1.PhaseAwaitingOperator, nil
+			return integreatlyv1alpha1.PhaseFailed, fmt.Errorf("error retrieving CSV: %v", err)
 		}
 
-		//if it's approved but not complete, then it's in progress
-		if ip.Status.Phase != operatorsv1alpha1.InstallPlanPhaseComplete && ip.Spec.Approved {
-			log.Infof("Install plan is not complete yet ", l.Fields{"install plan": target.Pkg})
-			return integreatlyv1alpha1.PhaseInProgress, nil
-			//if it's not approved by now, then it will not be approved by this version of the integreatly-operator
-		} else if !ip.Spec.Approved {
-			log.Infof("Upgrade installplan above the maximum allowed version", l.Fields{"install plan": target.Pkg})
+		if err := validateCSV(ipCSV); err != nil {
+			log.Warningf("CSV failed validation. Retrying operator installation", l.Fields{"error": err, "install plan": target.Pkg})
+			return retryInstallation(ctx, client, log, target, ipCSV, sub)
 		}
 	}
 
 	return integreatlyv1alpha1.PhaseCompleted, nil
 }
 
+func validateCSV(csv *operatorsv1alpha1.ClusterServiceVersion) error {
+	if csv.Spec.InstallStrategy.StrategyName == operatorsv1alpha1.InstallStrategyNameDeployment && len(csv.Spec.InstallStrategy.StrategySpec.DeploymentSpecs) == 0 {
+		return errors.New("No Deployment found in install strategy")
```

Comment:
Hey Sergio, not a big issue but I've found out that we should be trying to make sure that error messages begin with lower case letters rather than capitals